community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/AllStoreHolder.java

```diff
@@ -244,11 +244,8 @@ long graphPropertiesReference()
     @Override
     IndexReader indexReader( IndexReference index, boolean fresh ) throws IndexNotFoundKernelException
     {
-        SchemaIndexDescriptor schemaIndexDescriptor = index.isUnique() ?
-                                                      SchemaIndexDescriptorFactory.uniqueForLabel( index.label(), index.properties() ) :
-                                                      SchemaIndexDescriptorFactory.forLabel( index.label(), index.properties() );
-        return fresh ? storageReader.getFreshIndexReader( schemaIndexDescriptor ) :
-               storageReader.getIndexReader( schemaIndexDescriptor );
+        return fresh ? storageReader.getFreshIndexReader( (IndexDescriptor) index ) :
+               storageReader.getIndexReader( (IndexDescriptor) index );
```

Comment:
You'll get a `ClassCastException` (instead of e.g. `IndexNotFoundKernelException` if `index` is ever `IndexReference.NO_INDEX` for any reason. I think the code is too complicated and sprawling to guarantee that this will not happen.

Reply:
This might be a systemic problem. Could we change the code such that `IndexReference` instances are always valid, and use `Optional<IndexRerefence>`, or introduce an `OptionalIndexReference`, for the cases that may or may not have an index and where this question needs to be resolved before the data of the `IndexReference` can be accessed?

If we went with `OptionalIndexReference` as an interface, then `IndexReference` could extend it and resolving would just yield the same object instance, but with the resolved type.

Reply:
Aah, good point. Will explore using optional or maybe just checking for NO_INDEX under the hood before casting.

Made a card to address that as soon as this PR goes in, because it would be good to merge this ASAP to avoid conflicts and unlock other work.